kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs

```diff
@@ -246,9 +246,17 @@ impl<'tcx> GotocCtx<'tcx> {
             ($f:ident) => {{ codegen_intrinsic_binop!($f) }};
         }
 
-        // Intrinsics which encode a simple binary operation
-        macro_rules! codegen_intrinsic_boolean_binop {
-            ($f:ident) => {{ self.binop(p, fargs, |a, b| a.$f(b).cast_to(Type::c_bool())) }};
+        // Intrinsics which encode a pointer comparison (e.g., `ptr_guaranteed_eq`).
+        //
+        // Note(std): These intrinsics cannot be over-approximated because they
+        // are used for regular comparison in `std` when they shouldn't. An earlier
+        // version that returned nondet. values when the result of the comparison
```

Comment:
> because they are used for regular comparison in `std` when they shouldn't

This sounds a bit too strong for me. The documentation (https://doc.rust-lang.org/beta/std/primitive.pointer.html#method.guaranteed_eq) says that this may spuriously return `false` in **compile-time evaluation**, which I understood to mean that it may block some optimizations, etc., but it can never impact correctness since it behaves like `==`/`!=` at run-time (which is what we're modeling).

Reply:
Thanks for making this clear! Removed the comments here and in Kani docs.